Fix type confusion in BuilderConverter::convertPathOperation

chirags27 opened the pull request:
#### 1d37f36651144308e8d2c27a3cdabc939f3cf384
<pre>
Fix type confusion in BuilderConverter::convertPathOperation
<a href="https://bugs.webkit.org/show_bug.cgi?id=255735">https://bugs.webkit.org/show_bug.cgi?id=255735</a>
rdar://108253315

Reviewed by Chris Dumez.

This change fixes convertPathOperation so that it can deal with single
values instead of expecting a list of values towards the end.

* LayoutTests/fast/css/style-builder-type-confusion-crash-expected.txt: Added.
* LayoutTests/fast/css/style-builder-type-confusion-crash.html: Added.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPathOperation):

Canonical link: <a href="https://commits.webkit.org/263226@main">https://commits.webkit.org/263226@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43be92de113eb14ae46f90640964121267e7b92f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4053 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5399 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4211 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3936 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4141 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4028 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4113 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4008 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4194 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3573 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5292 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1696 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3548 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5622 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3608 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5021 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4010 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3225 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3527 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3548 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/973 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3577 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3807 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/db503e18eca5aabe708b3c8b64b87d477871229f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12981

chirags27 commented:
the changes here tweak the behavior in a subtle way which is causing test failures, I think I see why. Fixing it

cdumez commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -738,15 +738,26 @@ inline RefPtr<PathOperation> BuilderConverter::convertPathOperation(BuilderState
+    auto referenceBox = CSSBoxType::BoxMissing;
+    if (is<CSSValueList>(value)) {
+        for (auto& currentValue : downcast<CSSValueList>(value))
+            std::tie(operation, referenceBox) = processSingleValue(currentValue);
```
This will set both operation and referenceBox for each loop iteration, potentially overwriting the value from a previous iteration.

chirags27 replied:
Yes, fixed in latest commit

cdumez reviewed (approved):
Looks good now.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/43be92de113eb14ae46f90640964121267e7b92f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12981

chirags27 committed:
Fix type confusion in BuilderConverter::convertPathOperation
https://bugs.webkit.org/show_bug.cgi?id=255735
rdar://108253315

Reviewed by Chris Dumez.

This change fixes convertPathOperation so that it can deal with single
values instead of expecting a list of values towards the end.

* LayoutTests/fast/css/style-builder-type-confusion-crash-expected.txt: Added.
* LayoutTests/fast/css/style-builder-type-confusion-crash.html: Added.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPathOperation):

Canonical link: https://commits.webkit.org/263226@main

webkit-commit-queue commented:
Committed 263226@main (1d37f3665114): <https://commits.webkit.org/263226@main>

Reviewed commits have been landed. Closing PR #12981 and removing active labels.
